### PR TITLE
Improve diagnostics deduplication 2

### LINKED ISSUE #58207: Duplicate diagnostics
### 🔎 Search Terms

duplicate diagnostics, duplicate errors

### 🕗 Version & Regression Information

- This is the behavior in every version I tried

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.5.0-dev.20240415#code/JYOwLgpgTgZghgYwgAgMIHsC2AHANsBOSAEwBUBPbCAITgGcUBvAKGTeQG06AuZOsKKADmAXV4BBWgwDczAL7NQkWIhQYc+QiQpUAItGAA3CMWQt2nHnwHCxycfsHHishUujwk9qU1bs4vPyCIEKuiuAeqvaORiZmfmwARoE2IWEIuPR0yD7xFjDo6ACEvOp4BEQmOjT0EOnoIPzIAB7IALzIGVnIMc7IEM2QIMTZuebsBcWlWOVaVZQQvSZhuBBgyOiJAFa8jMgAzCnBQshy7WYHvABEYAAWUBAQV6eym1sczSLSQA

### 💻 Code

```ts
interface ComplicatedTypeBase {
    [s: string]: ABase;
}
interface ComplicatedTypeDerived {
    [s: string]: ADerived;
}
interface ABase {
    a: string;
}
interface ADerived {
    b: string;
}
class Base {
    foo!: ComplicatedTypeBase;
}
const x = class Derived extends Base {
    foo!: ComplicatedTypeDerived;
}
let obj: { 3: string } = { 3: "three" };
obj[x];
```

### 🙁 Actual behavior

Two instances of the same error on line `foo!: ComplicatedTypeDerived;`:
- `Property 'foo' in type 'Derived' is not assignable to the same property in base type 'Base'.
  Type 'ComplicatedTypeDerived' is not assignable to type 'ComplicatedTypeBase'.(2416)`
- `Property 'foo' in type 'Derived' is not assignable to the same property in base type 'Base'.
  Type 'ComplicatedTypeDerived' is not assignable to type 'ComplicatedTypeBase'.
    'string' index signatures are incompatible.
      Property 'a' is missing in type 'ADerived' but required in type 'ABase'.(2416)`

### 🙂 Expected behavior

A single instance of the error

### Additional information about the issue

_No response_

### PR DESCRIPTION
**Pull request by gabritto**
Fixes #58207.
Follow-up to #58220.

This applies deduplication for the two scenarios I'm aware of where we can produce different diagnostics (with different message and code) that report the same problem at the same location, depending on the current state of the type checker:
(1) not found diagnostics that may or may not have suggestions, and (2) type comparison diagnostics that may or may not suppress the top part of a diagnostic chain depending on whether we have a more useful elaboration or not.

Case (1) depends on the state of the checker because we have a constant limit on how many times we try to look for suggestion, for performance reasons. Case (2) depends on the state of the checker because we don't provide a more detailed elaboration when we already have a cached result for the type comparison.

The general idea here is to make these different diagnostics that refer to the same problem point to the same canonical diagnostic, and this canonical diagnostic is used for comparison and deciding equivalence.
e.g. in case (1), we can produce both "Cannot find name 'x'." and "Cannot find name 'x'. Did you mean 'y'?", so now the diagnostic with message "Cannot find name 'x'. Did you mean 'y'?" will have a canonical diagnostic of "Cannot find name 'x'.", and will be considered equivalent to "Cannot find name 'x'." (but the diagnostic with the suggestion will be preferred).

### CODE REVIEW

**Comment by gabritto** [comment]
@typescript-bot test it

**Comment by gabritto** [comment]
@typescript-bot perf test this

**Review by andrewbranch — APPROVED** [review]
Makes sense to me, though I would wait for another review from someone who knows more about `captureErrorCalculationState`/`resetErrorInfo`

File: tests/cases/compiler/duplicateErrorNameNotFound.ts
Comment (andrewbranch) [inline tests/cases/compiler/duplicateErrorNameNotFound.ts:7]: Did these tests actually produce duplicates, or were they just useful in getting near the code path for debugging?

File: tests/baselines/reference/mappedTypeRecursiveInference.types
Comment (weswigham) [inline tests/baselines/reference/mappedTypeRecursiveInference.types:7]: ...That's a big bump. Is this just caused by the extra `reportRelationError` call expanding the recursive type way more? Still feels like a lot more symbols.

File: src/compiler/utilities.ts
Comment (sheetalkamat) [inline src/compiler/utilities.ts:2175]: Does this need to be serialized in buildInfo when serializing diagnostic. Will we need this to deduplicate across multiple files?

File: tests/cases/compiler/duplicateErrorNameNotFound.ts
Comment (gabritto) [reply tests/cases/compiler/duplicateErrorNameNotFound.ts:7]: They produced duplicates, yes.

File: tests/baselines/reference/mappedTypeRecursiveInference.types
Comment (weswigham) [reply tests/baselines/reference/mappedTypeRecursiveInference.types:7]: This isn't necessarily something that aughta get fixed in this PR, but understanding where _so many symbols_ popped in from could be useful. It might indicate that we're not caching something we aughta be.

**Review by weswigham — APPROVED** [review]

**Comment by weswigham** [comment]
@typescript-bot perf test this faster

Juuuust in case

File: tests/baselines/reference/mappedTypeRecursiveInference.types
Comment (gabritto) [reply tests/baselines/reference/mappedTypeRecursiveInference.types:7]: > Is this just caused by the extra `reportRelationError` call expanding the recursive type way more?

I believe so. Since this test has an infinitely recurring type, it didn't seem like a great reason for concern to me, but I could be wrong.

File: src/compiler/utilities.ts
Comment (gabritto) [reply src/compiler/utilities.ts:2175]: > Will we need this to deduplicate across multiple files?

I don't think so, this should only be relevant for deduplicating diagnostics of a single file.

> Does this need to be serialized in buildInfo when serializing diagnostic?
What's the criterion for needing to serialize this in `buildInfo`?

**Comment by weswigham** [comment]
Minus 2% time in vscode? Neat. They must have had quite a few duplicates.

**Comment by jakebailey** [comment]
That result is right on the cusp of significance, and its before value is +/- 2%, so it's possible that it's not a real result, unfortunately. We're still p-hacking ourselves...

**Comment by gabritto** [comment]
> That result is right on the cusp of significance, and its before value is +/- 2%, so it's possible that it's not a real result, unfortunately. We're still p-hacking ourselves...

Yep, I'd suspect it's just noise. I don't expect anything to get faster, and I think nothing changed on extended tests, so doesn't seem like vscode has duplicate diagnostics. On the upside, nothing got significantly worse either :)

**Comment by weswigham** [comment]
> We're still p-hacking ourselves...

Should we un-fix `n` and instead fix `p`?

**Comment by jakebailey** [comment]
Running the vscode benchmark already takes half an hour, I don't see us being able to change `n` that much.

File: src/compiler/utilities.ts
Comment (sheetalkamat) [reply src/compiler/utilities.ts:2182]: you would need this info to deduplicate with errors from other files or errors emitted for same file in other phase (which obviously isn't true - like say same error is reported by declaration emit and semantic diagnostic just giving hypothetical example here)

Comment (sheetalkamat) [reply src/compiler/utilities.ts:2182]: Given this doesnt get used beyond file boundaries this should be ok to be not serialized. 
That is checking files in any order will result in same error in that file (which we were anyways assuming till now) this is good.

**Review by sheetalkamat — APPROVED** [review]

**Pull request merged**
